conformance/src/main/scala/za/co/absa/enceladus/conformance/interpreter/DynamicInterpreter.scala

```diff
@@ -214,43 +213,20 @@ object DynamicInterpreter {
     * @return The list of lists of conformance rule groups
     */
   private def groupMappingRules(rules: List[ConformanceRule], schema: StructType): List[List[ConformanceRule]] = {
-    val groups = new ListBuffer[List[ConformanceRule]]
-    var group = new ListBuffer[ConformanceRule]
-    var lastArrayParent = ""
-
-    def pushGroup(arrayPath: String = ""): Unit = {
-      if (group.nonEmpty) {
-        groups += group.toList
-        group = new ListBuffer[ConformanceRule]
-      }
-      lastArrayParent = arrayPath
-    }
-
-    for (rule <- rules) {
-      rule match {
-        case m: MappingConformanceRule =>
-          val arrayParent = SchemaUtils.getDeepestArrayPath(schema, rule.outputColumn)
-          arrayParent match {
-            case Some(arr) =>
-              if (lastArrayParent != arr) {
-                pushGroup(arr)
-              }
-              group += m
-            case None =>
-              pushGroup()
-              groups += List(m)
-          }
-
-          // Something
-        case a =>
-          pushGroup()
-          groups += List(a)
-      }
-    }
-    pushGroup()
-    groups.toList
+    Algorithms.stableGroupByOption[ConformanceRule, String](rules, {
+      case m: MappingConformanceRule => SchemaUtils.getDeepestArrayPath(schema, m.outputColumn)
+      case _ => None
+    }).map(_.toList).toList
   }
 
+  /**
+    * Adds an explosion pseudo-rule before each group of more than 1 mapping rule.
+    * Adds a collection pseudo-rule after each group of more than 1 mapping rule
+    *
+    * @param ruleGroups A list of rules grouped together if they are mapping rules operating at the same array level
+    * @param schema     A schema of a dataset
+    * @return The new list of rules that might contain pseudo-rules
+    */
```

Comment:
We talked that private API doesn't have to be commented, but I really appreciate you taking time and doing so.

Reply:
Thanks!